server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -5786,6 +5788,69 @@ private void runTestDeleteFailure(final CheckedBiConsumer<InternalEngine, Engine
         }
     }
 
+    public void testIndexThrottling() throws Exception {
+        engine = spy(engine);
+
+        long sleepInMillis = 1000;
+        engine.activateThrottling();
+        Long timeTaken = concurrentIndexingWithSleep(sleepInMillis);
+        assertTrue(timeTaken > (2 * sleepInMillis));
+
+        engine.deactivateThrottling();
+        timeTaken = concurrentIndexingWithSleep(sleepInMillis);
+        assertTrue(timeTaken < (2 * sleepInMillis));
+    }
+
+    private long concurrentIndexingWithSleep(long sleepInMillis) throws Exception {
+        CountDownLatch countDownLatch = new CountDownLatch(3);
+        Mockito.doAnswer((Answer<InternalEngine.IndexingStrategy>) invocation -> {
+            try {
+                Thread.sleep(sleepInMillis);
```

Comment:
I know it's not a good design but couldn't think of a better way to ensure sequential indexing in case of throttling